qdl.c

```diff
@@ -193,44 +195,70 @@ static int decode_programmer_archive(struct sahara_image *images)
 		ptr += sizeof(*hdr);
 		if (ptr + namesize > end || ptr + filesize + namesize > end) {
 			ux_err("programmer archive is truncated\n");
-			return -1;
+			break;
 		}
 
 		if (namesize > sizeof(name)) {
 			ux_err("unexpected filename length in progammer archive\n");
-			return -1;
+			break;
 		}
 		memcpy(name, ptr, namesize);
 
-		if (!memcmp(name, "TRAILER!!!", 11))
+		if (!memcmp(name, "TRAILER!!!", 11)) {
+			ret = 1;
 			break;
+		}
 
 		tok = strtok_r(name, ":", &save);
 		id = strtoul(tok, NULL, 0);
 		if (id == 0 || id >= MAPPING_SZ) {
 			ux_err("invalid image id \"%s\" in programmer archive\n", tok);
-			return -1;
+			break;
 		}
 
 		ptr += namesize;
 		ptr = ALIGN_UP(ptr, 4);
 
 		tok = strtok_r(NULL, ":", &save);
-		if (tok)
+		if (tok) {
 			images[id].name = strdup(tok);
+			if (!images[id].name) {
+				ux_err("failed to allocated buffer for an image name\n");
+				break;
+			}
+		}
 		images[id].len = filesize;
 		images[id].ptr = malloc(filesize);
+		if (!images[id].ptr) {
+			ux_err("failed to allocated buffer for an image, len = %z\n", images[id].len);
+			break;
+		}
 		memcpy(images[id].ptr, ptr, filesize);
 
 		ptr += filesize;
 		ptr = ALIGN_UP(ptr, 4);
 	}
 
-	free(blob->ptr);
-	blob->ptr = NULL;
-	blob->len = 0;
+	/*
+	 * even if some images aren't initialized, it's safe to enumerate over each
+	 * image and try to free, as otherwise image[i] will be filled with zeros.
```

Comment:
If needed, this would better be added to the comment above the function, stating that the images array must be zero-initialized (but I don't think it's necessary to document that).